### PR TITLE
Uma opcodes

### PR DESCRIPTION
**Pull request by gianbelinche**
Closes https://github.com/lambdaclass/era_vm/issues/12
This PR add the opcodes related to UMA, `ld, st, ld.1, ld.2, st.1, st.2` as well as `.inc`variants
It builds on top of https://github.com/lambdaclass/era_vm/pull/30 so that one needs to be merged first

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -96,7 +102,15 @@ pub fn run_program_with_custom_state(bin_path: &str, mut vm: VMState) -> (U256,
...
                     break;
                 }
-                Variant::UMA(_) => todo!(),
+                Variant::UMA(uma_variant) => match uma_variant {
+                    UMAOpcode::HeapRead => _heap_read(&mut vm, &opcode),
+                    UMAOpcode::HeapWrite => _heap_write(&mut vm, &opcode),
+                    UMAOpcode::AuxHeapRead => _aux_heap_read(&mut vm, &opcode),
+                    UMAOpcode::AuxHeapWrite => _aux_heap_write(&mut vm, &opcode),
+                    UMAOpcode::FatPointerRead => _fat_pointer_read(&mut vm, &opcode),
+                    UMAOpcode::StaticMemoryRead => todo!(),
```
Comment (jorbush) [inline src/lib.rs:111]: In issue #12, include the `StaticMemoryRead` and `StaticMemoryWrite` opcodes. 
I assume you will do this in another PR, but just to confirm.

Comment (gianbelinche) [reply src/lib.rs:111]: In the VM specification there is no mention of StaticMemoryRead or StaticMemoryWrite, so they won't be implemented.
Also, on the vm2, it is marked as an unimplemented instruction.

File: src/state.rs [hunk 2]
```diff
@@ -250,3 +261,49 @@ impl Stack {
...
+            let old_size = self.heap.len() as u32;
+            self.heap.resize(address as usize + 1, 0);
+            return MEMORY_GROWTH_ERGS_PER_BYTE * (address - old_size + 1);
+        }
+        0
+    }
+
+    pub fn store(&mut self, address: u32, value: U256) {
+        for i in 0..32 {
+            self.heap[address as usize + i] = value.byte(31 - i);
```
Comment (fkrause98) [inline src/state.rs:287]: Any reason why we aren't using .to_big_endian and simply extending the vector with it? The spec states this is interpreted as big endian. Page 189, if you need further reference.

File: src/state.rs [hunk 3]
```diff
@@ -250,3 +261,49 @@ impl Stack {
...
+        }
+    }
+
+    pub fn read(&mut self, address: u32) -> U256 {
+        let mut result = U256::zero();
+        for i in 0..32 {
```
Comment (fkrause98) [inline src/state.rs:293]: Here you can also use to_big_endian, but I think that would be overkill here, and what you're doing is more clear, so I would leave it as is.

**Review by fkrause98 — COMMENTED** [review]
Looks good, let's discuss the comments I've made before merging.

Great job with the generated tests and the feature.🚀

File: src/op_handlers/heap_read.rs [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+use crate::value::TaggedValue;
+use crate::{opcode::Opcode, state::VMState};
+
+pub fn _heap_read(vm: &mut VMState, opcode: &Opcode) {
+    let (src0, _) = address_operands_read(vm, opcode);
+    if src0.is_pointer {
+        panic!("Invalid operands for heap_read");
+    }
+
+    if src0.value > U256::from(MAX_OFFSET_TO_DEREF_LOW_U32) {
```
[the remark this answers is not in the document]
Comment (gianbelinche) [reply src/op_handlers/heap_read.rs:14]: I think its better to leave it as is, first because adding an extra function for 4-5 lines seems too much, even if they repeat, and second because this is going to eventually change when we add proper error handling, and then we will implement it in the best way possible

File: src/state.rs (hunk 2 above)
Comment (gianbelinche) [reply src/state.rs:287]: I didn't realize that, however, the best thing I can come up with is
```
let mut bytes: [u8;32] = [0;32];
value.to_big_endian(&mut bytes);
for i in 0..32 {
      self.heap[address as usize + i] = bytes[i];
}
```
I can't just extend the vector, since there is an address were it should be stored

File: src/state.rs (hunk 3 above)
Comment (gianbelinche) [reply src/state.rs:293]: Perfect, I leave it as is

File: src/op_handlers/aux_heap_read.rs [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+    if src0.is_pointer {
+        panic!("Invalid operands for heap_read");
+    }
+
+    if src0.value > U256::from(MAX_OFFSET_TO_DEREF_LOW_U32) {
+        panic!("Address too large for heap_read");
+    }
+    let addr = src0.value.low_u32();
+
+    vm.current_frame.aux_heap.expand_memory(addr + 32); // TODO: Handle ergs cost
```
[the remark this answers is not in the document]
Comment (gianbelinche) [reply src/op_handlers/aux_heap_read.rs]: Done!

**Review by fkrause98 — APPROVED** [review]
Great job!

File: src/state.rs (hunk 2 above)
Comment (fkrause98) [reply src/state.rs]: Yup

**Review by fkrause98 — APPROVED** [review]

**Review by juan518munoz — APPROVED** [review]
LGTM
